### PR TITLE
Remove gateway Linking

### PR DESCRIPTION
**Pull request by sarvesh-ost**
Following features are implemented with this PR: 

1. Removed gateway linking logic. 
2. Deleted Gateway.sol and CoGateway.sol
3. Added gateway activation logic. 
4.  Unit test cases of the constructor of GatewayBase.sol, EIP20Gateway.sol, EIP20CoGateway.sol and activate gateway logic. 

#490

### CODE REVIEW

File: contracts/gateway/EIP20CoGateway.sol [hunk 1]
```diff
@@ -52,18 +54,15 @@ RevertRedemptionIntentConfirmed   --->   revertRedemption
         progressUnstakeWithProof  <---   progressRedemptionWithProof
 -------------------------------------------------------------------------------
 */
-
-import "./CoGateway.sol";
-
+import "./UtilityTokenInterface.sol";
```
Comment (schemar) [inline contracts/gateway/EIP20CoGateway.sol]: Shouldn't all the imports be next to each other? 🤔

File: contracts/gateway/EIP20CoGateway.sol [hunk 2]
```diff
@@ -179,6 +178,12 @@ contract EIP20CoGateway is CoGateway {
 
     /* public variables */
 
+    /** address of utility token. */
+    address public utilityToken;
+
+    /** address of value token. */
+    address public valueToken;
```
Comment (schemar) [inline contracts/gateway/EIP20CoGateway.sol]: Why not of type interface? Do you use the imported `UtilityTokenInterface` anywhere?

File: contracts/gateway/EIP20Gateway.sol [hunk 3]
```diff
@@ -964,7 +980,56 @@ contract EIP20Gateway is Gateway {
...
+        );
+
+        remoteGateway = _coGatewayAddress;
+
+        // update the encodedGatewayPath
+        encodedGatewayPath = GatewayLib.bytes32ToBytes(
+            keccak256(abi.encodePacked(remoteGateway))
+        );
+        activated = true;
+        return true;
```
Comment (schemar) [inline contracts/gateway/EIP20Gateway.sol]: How about named return values?

File: contracts/gateway/EIP20Gateway.sol [hunk 4]
```diff
@@ -964,7 +980,56 @@ contract EIP20Gateway is Gateway {
...
+    *         Organisation address only once by passing co-gateway address.
+    *
+    * @param _coGatewayAddress Address of cogateway.
+    *
+    * @return `true` if value is set
+    */
+    function activateGateway(address _coGatewayAddress)
+        external
+        onlyOrganisation
+        returns (bool)
```
Comment (schemar) [inline contracts/gateway/EIP20Gateway.sol]: Not sure, but I think if you have a multi-line function head, the arguments must also be multi-line.

File: contracts/gateway/EIP20Gateway.sol [hunk 5]
```diff
@@ -964,7 +980,56 @@ contract EIP20Gateway is Gateway {
...
+    {
+
+        require(
+            remoteGateway == address(0),
+            'Gateway was already activated once.'
+        );
+
+        remoteGateway = _coGatewayAddress;
```
Comment (schemar) [inline contracts/gateway/EIP20Gateway.sol]: Why not `coGateway` in storage or `_remoteGateway` as argument?

File: contracts/gateway/EIP20Gateway.sol [hunk 6]
```diff
@@ -964,7 +980,56 @@ contract EIP20Gateway is Gateway {
...
+    */
+    function activateGateway(address _coGatewayAddress)
+        external
+        onlyOrganisation
+        returns (bool)
+    {
+
+        require(
+            remoteGateway == address(0),
+            'Gateway was already activated once.'
```
Comment (schemar) [inline contracts/gateway/EIP20Gateway.sol]: double quotes in solidity

File: contracts/gateway/GatewayBase.sol [hunk 7]
```diff
@@ -150,10 +138,10 @@ contract GatewayBase {
         _;
     }
 
-    /** checks that contract is linked and is not deactivated */
+    /** checks that contract is activated */
     modifier isActive() {
         require(
-            deactivated == false && linked == true,
+            activated == true,
             "Contract is restricted to use"
```
Comment (schemar) [inline contracts/gateway/GatewayBase.sol]: Can you improve the message while you're here?

File: contracts/gateway/MessageBus.sol [hunk 8]
```diff
@@ -61,10 +61,6 @@ library MessageBus {
     /** Message */
     struct Message {
 
-        /**
-         * intent hash, this can be staking intent hash, redemption intent
-         * hash, gateway linking hash
-         */
```
Comment (schemar) [inline contracts/gateway/MessageBus.sol]: Why removed?

File: test/gateway/eip20_cogateway/constructor.js [hunk 9]
```diff
@@ -0,0 +1,176 @@
+const CoGateway = artifacts.require("EIP20CoGateway");
```
Comment (schemar) [inline test/gateway/eip20_cogateway/constructor.js]: Missing license

File: test/gateway/eip20_cogateway/constructor.js [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+MockToken = artifacts.require("MockToken");
```
Comment (schemar) [inline test/gateway/eip20_cogateway/constructor.js]:
```suggestion
const MockToken = artifacts.require("MockToken");
```

File: test/gateway/eip20_cogateway/constructor.js [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+
+const Utils = require("./../../test_lib/utils"),
+    BN = require('bn.js');
+
+const NullAddress = "0x0000000000000000000000000000000000000000";
+
+contract('EIP20CoGateway ', function (accounts) {
+
+
+    describe('constructor', async function () {
```
Comment (schemar) [inline test/gateway/eip20_cogateway/constructor.js]: We don't use `describe` at the moment. Instead, it should be `contract('EIP20CoGateway.constructor()'`, at least until we decide a better way.

File: test/gateway/eip20_cogateway/constructor.js [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+                    organisation,
+                    gatewayAddress
+                );
+
+            let valueTokenAddress = await coGateway.valueToken.call();
+
+            assert.equal(
+                valueTokenAddress,
+                valueToken.address,
+                "Invalid valueTokenAddress address from contract."
```
Comment (schemar) [inline test/gateway/eip20_cogateway/constructor.js]: single quotes in JavaScript. I think.

File: test/gateway/eip20_gateway/activate_gateway.js [hunk 13]
```diff
@@ -0,0 +1,92 @@
+const Gateway = artifacts.require("./EIP20Gateway.sol")
+    , BN = require('bn.js');
+
+const Utils = require('../../../test/test_lib/utils');
+
+
+contract('GatewayBase.sol', function (accounts) {
+
+    describe('Deactivate gateway', async () => {
```
Comment (schemar) [inline test/gateway/eip20_gateway/activate_gateway.js]: Same as above. Plus: wrong name `GatewayBase`. Plus: deactivate should be in a different file that tests the deactivate method.

File: test/gateway/eip20_gateway/activate_gateway.js [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+
+        it('should not deactivate if already deactivated', async function () {
+
+            await gateway.deactivateGateway({from: organisation});
+            await Utils.expectThrow(gateway.deactivateGateway.call({from: organisation}));
+        });
+
+        it('should deactivated by organization only', async function () {
+
+            await Utils.expectThrow(gateway.deactivateGateway.call({from: accounts[0]}));
```
Comment (schemar) [inline test/gateway/eip20_gateway/activate_gateway.js]: You should check for the correct revert error messages.

File: test/gateway/eip20_gateway/constructor.js [hunk 15]
```diff
@@ -0,0 +1,175 @@
...
+
+const Utils = require("./../../test_lib/utils"),
+    BN = require('bn.js');
+
+const NullAddress = "0x0000000000000000000000000000000000000000";
+
+contract('EIP20Gateway ', function (accounts) {
+
+
+    describe('constructor', async function () {
```
Comment (schemar) [inline test/gateway/eip20_gateway/constructor.js]: see above

File: test/gateway/eip20_gateway/constructor.js [hunk 16]
```diff
@@ -0,0 +1,175 @@
+const Gateway = artifacts.require("EIP20Gateway");
+MockToken = artifacts.require("MockToken");
```
Comment (schemar) [inline test/gateway/eip20_gateway/constructor.js]:
```suggestion
const MockToken = artifacts.require("MockToken");
```

File: test/gateway/eip20_gateway/constructor.js [hunk 17]
```diff
@@ -0,0 +1,175 @@
...
+                );
+
+            assert(
+                web3.utils.isAddress(gateway.address),
+                "Returned value is not a valid address."
+            );
+        });
+
+    });
+});
```
Comment (schemar) [inline test/gateway/eip20_gateway/constructor.js]:
```suggestion
});

```

File: test/gateway/eip20_gateway/constructor.js [hunk 18]
```diff
@@ -0,0 +1,175 @@
+const Gateway = artifacts.require("EIP20Gateway");
```
Comment (schemar) [inline test/gateway/eip20_gateway/constructor.js]: More files are missing licenses

File: test/gateway/eip20_gateway/stake.js [hunk 19]
```diff
@@ -0,0 +1,324 @@
...
+    gatewayTest = new EIP20GatewayKlass(gateway, mockToken, baseToken);
+
+
+    hashLockObj = utils.generateHashLock();
+
+    facilitator = accounts[4];
+    nonce = await  helper.getNonce(accounts[1]);
+    stakeAmount = new BN(100000000000);
+    beneficiary = accounts[2];
+    stakerAddress = accounts[1];
```
Comment (schemar) [inline test/gateway/eip20_gateway/stake.js]: Same as organisation and core. On purpose?

File: test/gateway/eip20_gateway/stake.js [hunk 20]
```diff
@@ -0,0 +1,324 @@
...
+        params,
+        resultType,
+        expectedResult,
+        txOption
+    );
+}
+
+contract('EIP20Gateway ', function (accounts) {
+
+    describe('stake', async function () {
```
Comment (schemar) [inline test/gateway/eip20_gateway/stake.js]: see above

File: contracts/gateway/EIP20Gateway.sol (hunk 5 above)
Comment (sarvesh-ost) [reply contracts/gateway/EIP20Gateway.sol]: remoteGateway storage is defined in `GatewayBase`, which can be both gateway and co-gateway. 
But here `activateGateway` method is defined in `EIP20Gateway` context which means remote gateway is know here which is `Co-gateway`.

File: contracts/gateway/MessageBus.sol (hunk 8 above)
Comment (sarvesh-ost) [reply contracts/gateway/MessageBus.sol]: Message bus shouldn't know about a different kind of intent hash(Message bus). I feel there shouldn't be any mention of the stake, redemption etc in the message bus.

File: test/gateway/eip20_gateway/stake.js (hunk 19 above)
Comment (sarvesh-ost) [reply test/gateway/eip20_gateway/stake.js]: No, I didn't make any logic change in this file.  Though I have changed organisation and core addresses.

File: test/gateway/eip20_gateway/activate_gateway.js (hunk 13 above)
Comment (sarvesh-ost) [reply test/gateway/eip20_gateway/activate_gateway.js]: All the feedback around `activate_gateway.js` is done in pr #518 
Please check [this](https://github.com/OpenSTFoundation/mosaic-contracts/pull/518/commits/51e49539049b83be83e5156fb1ab228da580da94) commit.

File: contracts/gateway/GatewayBase.sol [hunk 21]
```diff
@@ -165,40 +153,28 @@ contract GatewayBase {
...
      *                stake process.
      * @param _organisation Organisation address.
      */
     constructor(
         CoreInterface _core,
-        address _messageBus,
         uint256 _bounty,
         address _organisation
     )
     public
```
Comment (deepesh-kn) [inline contracts/gateway/GatewayBase.sol]: Indentation

File: contracts/test/TestEIP20Gateway.sol [hunk 22]
```diff
@@ -1,27 +1,27 @@
 pragma solidity ^0.5.0;
 
 import "../gateway/EIP20Gateway.sol";
+import "../gateway/MessageBus.sol";
 
-contract MockEIP20Gateway is EIP20Gateway {
+contract TestEIP20Gateway is EIP20Gateway {
 
+    MessageBus messageBus;
```
Comment (deepesh-kn) [inline contracts/test/TestEIP20Gateway.sol]: New line after this statement

File: contracts/test/TestEIP20Gateway.sol [hunk 23]
```diff
@@ -1,27 +1,27 @@
 pragma solidity ^0.5.0;
 
 import "../gateway/EIP20Gateway.sol";
+import "../gateway/MessageBus.sol";
```
Comment (deepesh-kn) [inline contracts/test/TestEIP20Gateway.sol]: Copyright documentations are missing.

File: test/gateway/eip20_cogateway/constructor.js [hunk 24]
```diff
@@ -0,0 +1,192 @@
...
+                utilityToken.address,
+                coreAddress,
+                bountyAmount,
+                organisation,
+                gatewayAddress
+            );
+
+        let valueTokenAddress = await coGateway.valueToken.call();
+
+        assert.equal(
```
Comment (deepesh-kn) [inline test/gateway/eip20_cogateway/constructor.js]: We can change to `strictEqual` at multiple places in this file.

File: test/gateway/eip20_gateway/activate_gateway.js [hunk 25]
```diff
@@ -0,0 +1,92 @@
...
+            );
+
+            let coGateway = accounts[5];
+            await  gateway.activateGateway(coGateway, {from: organisation});
+
+        });
+
+        it('should deactivate if activated', async function () {
+
+            assert((await gateway.deactivateGateway.call({from: organisation})));
```
Comment (deepesh-kn) [inline test/gateway/eip20_gateway/activate_gateway.js]: We can check `activated` variable value.

File: test/gateway/gateway_base/construction.js [hunk 26]
```diff
@@ -3,35 +3,31 @@ const GatewayBase = artifacts.require("./GatewayBase.sol")
...
       let gatewayBaseInstance = await GatewayBase.new(
         core,
-        messageBus,
         bounty,
         organisation
       );
 
-      assert.equal(core, await gatewayBaseInstance.core.call());
-      assert.equal(messageBus, await gatewayBaseInstance.messageBus.call());
+      assert.equal(core,await gatewayBaseInstance.core.call());
```
Comment (deepesh-kn) [inline test/gateway/gateway_base/construction.js]: We should have a reason for an assert.

**Pull request merged**